### PR TITLE
Adding `CONVERT_TZ` and `DATETIME` functions to SQL and PPL 

### PR DESCRIPTION
**Pull request by MitchellGale-BitQuill**
### Description
PR adds support for convert_tz and Datetime for PPL and SQL languages.

### convert_tz

Syntax for convert_tz:
```convert_tz(datetime string, from timezone, to timezone)```
It converts the `datetime string` from the `from timezone` to the `to timezone`. Converts from negative to positive timezone, from positive to negative, from positive to positive or negative to negative time zones. Works across year roll-over and leap years.
Example:
```
opensearchsql> SELECT convert_tz('2021-05-12 00:00:00','-12:00','+13:00'); 
fetched rows / total rows = 1/1
+-------------------------------------------------------+
| convert_tz('2021-05-12 00:00:00','-12:00','+13:00')   |
|-------------------------------------------------------|
| 2021-05-13 01:00:00                                   |
+-------------------------------------------------------+
``` 

```
opensearchsql> SELECT convert_tz('2021-05-12 12:34:56','-00:00','+01:00');
fetched rows / total rows = 1/1
+-------------------------------------------------------+
| convert_tz('2021-05-12 12:34:56','-00:00','+01:00')   |
|-------------------------------------------------------|
| 2021-05-12 13:34:56                                   |
+-------------------------------------------------------+
```

#### nulls
It returns null for time zones outside of the allowed range (allow range is -13:59 -> +14:00)
```
opensearchsql> SELECT convert_tz('2021-05-12 12:34:56','-00:00','+15:00');
fetched rows / total rows = 1/1
+-------------------------------------------------------+
| convert_tz('2021-05-12 12:34:56','-00:00','+14:01')   |
|-------------------------------------------------------|
| null                                                  |
+-------------------------------------------------------+
```

```
opensearchsql> SELECT convert_tz('2021-05-12 12:34:56','-13:00','+10:00');
fetched rows / total rows = 1/1
+-------------------------------------------------------+
| convert_tz('2021-05-12 12:34:56','-14:00','+10:00')   |
|-------------------------------------------------------|
| null                                                  |
+-------------------------------------------------------+
```

#### Leap year

```
opensearchsql> SELECT convert_tz('2005-02-28 20:00:00','-10:00','+10:00');
fetched rows / total rows = 1/1
+-------------------------------------------------------+
| convert_tz('2005-02-28 20:00:00','-10:00','+10:00')   |
|-------------------------------------------------------|
| 2005-03-01 16:00:00                                   |
+-------------------------------------------------------+
```

```
opensearchsql> SELECT convert_tz('2004-02-28 20:00:00','-10:00','+10:00');
fetched rows / total rows = 1/1
+-------------------------------------------------------+
| convert_tz('2004-02-28 20:00:00','-10:00','+10:00')   |
|-------------------------------------------------------|
| 2004-02-29 16:00:00                                   |
+-------------------------------------------------------+
```

### DateTime
```DATETIME(timestamp_expression [, time_zone])```

```opensearchsql> SELECT DATETIME('2008-12-25 05:30:00-05:00', '+05:00'); 
fetched rows / total rows = 1/1
+---------------------------------------------------+
| DATETIME('2008-12-25 05:30:00-05:00', '+05:00')   |
|---------------------------------------------------|
| 2008-12-25 15:30:00                               |
+---------------------------------------------------+
```

```
opensearchsql> SELECT DATETIME('2008-12-25 05:30:00', '+05:00');
fetched rows / total rows = 1/1
+---------------------------------------------+
| DATETIME('2008-12-25 05:30:00', '+05:00')   |
|---------------------------------------------|
| 2008-12-25 10:30:00                         |
+---------------------------------------------+
```

```
opensearchsql> SELECT DATETIME('2008-12-25 05:30:00+10:00'); 
fetched rows / total rows = 1/1
+-----------------------------------------+
| DATETIME('2008-12-25 05:30:00+10:00')   |
|-----------------------------------------|
| 2008-12-25 05:30:00                     |
+-----------------------------------------+
```

```
opensearchsql> SELECT DATETIME('2008-12-25 05:30:00');
fetched rows / total rows = 1/1
+-----------------------------------+
| DATETIME('2008-12-25 05:30:00')   |
|-----------------------------------|
| 2008-12-25 05:30:00               |
+-----------------------------------+

```

### Issues Resolved
[#703](https://github.com/opendistro-for-elasticsearch/sql/issues/709)
[#46 ](https://github.com/opensearch-project/sql/issues/46)
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: docs/user/dql/functions.rst [hunk 1]
```diff
@@ -893,6 +982,100 @@ Example::
...
+DATETIME, STRING -> DATETIME
+
+DATETIME -> DATETIME
+
+Example::
+
+    os> SELECT DATETIME('2008-12-25 05:30:00+00:00', 'America/Los_Angeles')
+    fetched rows / total rows = 1/1
+    +----------------------------------------------------------------+
+    | DATETIME('2008-12-25 05:30:00+00:00', 'America/Los_Angeles')   |
```
Comment (dai-chen) [inline docs/user/dql/functions.rst:1006]: Just wondering does `convert_tz` support timezone name as well? Didn't see example above.

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 2]
```diff
@@ -468,6 +548,61 @@ private ExprValue exprDate(ExprValue exprValue) {
...
+   * DateTime implementation for ExprValue.
+   *
+   * @param dateTime ExprValue of String type.
+   * @param timeZone ExprValue of String type (or null).
+   * @return ExprValue of date type.
+   */
+  private ExprValue exprDateTime(ExprValue dateTime, ExprValue timeZone) {
+    String defaultTimeZone = TimeZone.getDefault().getID();
+    DateTimeFormatter formatDT = DateTimeFormatter.ofPattern("uuuu-MM-dd HH:mm:ss[xxx]")
+        .withResolverStyle(ResolverStyle.STRICT);
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:561]: Please check this comment and see if any change needed here: https://github.com/opensearch-project/sql/pull/835#discussion_r973371510

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 3]
```diff
@@ -48,6 +57,9 @@
  */
 @UtilityClass
 public class DateTimeFunction {
+  String timeZoneMax = "+14:00";
+  String timeZoneMin = "-13:59";
+  String timeZoneZero = "+00:00";
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:62]: Maybe better to move this along with `isValidMySqlTimeZoneId` to some time zone utility class?

**Comment by dai-chen** [comment]
Could you point me to the documentation for `DATETIME` function? For some reason I didn't find it in MySQL...

**Comment by MitchellGale-BitQuill** [comment]
@dai-chen `DATETIME` doesn't exist in MySQL, but it's implemented like a constructor, similar to the `TIMESTATMP` but for datetime.
Also listed here [https://github.com/opensearch-project/sql/issues/722](https://github.com/opensearch-project/sql/issues/722).

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 3 above)
Comment (dai-chen) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:62]: Similar thought for your reference: https://github.com/opensearch-project/sql/pull/835#discussion_r974804923

Comment (MitchellGale-BitQuill) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java]: Fixed in abae9618

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 2 above)
Comment (MitchellGale-BitQuill) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java]: Fixed in eb0c89498df5722ec494ddd9eac0554f65f77007

File: docs/user/dql/functions.rst (hunk 1 above)
Comment (MitchellGale-BitQuill) [reply docs/user/dql/functions.rst:1015]: Fixed in 1fbed3f609b91568ec6251064e8bbdeb9eed83b5

**Review by Yury-Fridlyand — APPROVED** [review]

**Review by acarbonetto — APPROVED** [review]

**Pull request merged**